File: scripts/bird_species.py
import torch
import torch.nn as nn
import torch.optim as optim
from torch.optim import lr_scheduler
import numpy as np
import torchvision
from torchvision import datasets, models, transforms

import matplotlib.pyplot as plt
import time
import os
import copy

data_transforms = {
    'train': transforms.Compose([
        transforms.RandomResizedCrop(224),
        transforms.ColorJitter(),
        transforms.RandomHorizontalFlip(),
        transforms.ToTensor(),
        transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
    ]),
    'valid': transforms.Compose([
        transforms.Resize(256),
        transforms.CenterCrop(224),
        transforms.ToTensor(),
        transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
    ]),
    'test': transforms.Compose([
        transforms.Resize(256),
        transforms.CenterCrop(224),
        transforms.ToTensor(),
        transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
    ]),
}

# https://www.kaggle.com/gpiosenka/100-bird-species/data
data_dir = 'data/bird_species'

image_datasets = {x: datasets.ImageFolder(os.path.join(data_dir, x),
                                          data_transforms[x])
                  for x in ['train', 'valid', 'test']}
image_datasets

dataloaders = {x: torch.utils.data.DataLoader(image_datasets[x], batch_size=16,
                                             shuffle=True, num_workers=8)
              for x in ['train', 'valid', 'test']}
dataloaders

dataset_sizes = {x: len(image_datasets[x]) for x in ['train', 'valid', 'test']}
dataset_sizes

class_names = image_datasets['train'].classes
class_names

device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

def imshow(inp, title=None):
    """Imshow for Tensor."""
    inp = inp.numpy().transpose((1, 2, 0))
    mean = np.array([0.485, 0.456, 0.406])
    std = np.array([0.229, 0.224, 0.225])
    inp = std * inp + mean
    inp = np.clip(inp, 0, 1)
    plt.imshow(inp)
    if title is not None:
        plt.title(title)
    plt.pause(0.001)  

inputs, classes = next(iter(dataloaders['train']))

out = torchvision.utils.make_grid(inputs)

imshow(out, title=[class_names[x] for x in classes])

def train_model(model, criterion, optimizer, scheduler, num_epochs=25):
    since = time.time()
    best_model_wts = copy.deepcopy(model.state_dict())
    best_acc = 0.0
    for epoch in range(num_epochs):
        print('Epoch {}/{}'.format(epoch, num_epochs - 1))
        print('-' * 10)
        for phase in ['train', 'valid']:
            if phase == 'train':
                model.train() 
            else:
                model.eval()   
            running_loss = 0.0
            running_corrects = 0
            for inputs, labels in dataloaders[phase]:
                inputs = inputs.to(device)
                labels = labels.to(device)
                optimizer.zero_grad()
                with torch.set_grad_enabled(phase == 'train'):
                    outputs = model(inputs)
                    _, preds = torch.max(outputs, 1)
                    loss = criterion(outputs, labels)
                    if phase == 'train':
                        loss.backward()
                        optimizer.step()
                running_loss += loss.item() * inputs.size(0)
                running_corrects += torch.sum(preds == labels.data)
            if phase == 'train':
                scheduler.step()
            epoch_loss = running_loss / dataset_sizes[phase]
            epoch_acc = running_corrects.double() / dataset_sizes[phase]
            print('{} Loss: {:.4f} Acc: {:.4f}'.format(
                phase, epoch_loss, epoch_acc))
            if phase == 'valid' and epoch_acc > best_acc:
                best_acc = epoch_acc
                best_model_wts = copy.deepcopy(model.state_dict())
        print()
    time_elapsed = time.time() - since
    print('Training complete in {:.0f}m {:.0f}s'.format(
        time_elapsed // 60, time_elapsed % 60))
    print('Best val Acc: {:4f}'.format(best_acc))
    model.load_state_dict(best_model_wts)
    return model



model = models.resnet18(pretrained=True)
model

for param in model.parameters():
    param.requires_grad = False

num_features = model.fc.in_features

model.fc = nn.Linear(num_features, len(class_names))

model = model.to(device)

criterion = nn.CrossEntropyLoss()

optimizer = optim.SGD(model.parameters(), lr=0.001, momentum=0.9)

exp_lr_scheduler = lr_scheduler.StepLR(optimizer, step_size=7, gamma=0.1)

model = train_model(model, criterion, optimizer, exp_lr_scheduler,
                       num_epochs=1)

phase = 'test'
with torch.no_grad():
        running_num_correct = 0
        for inputs, labels in dataloaders[phase]:
            inputs = inputs.to(device)
            labels = labels.to(device)
            outputs = model(inputs)
            _, preds = torch.max(outputs, 1)
            running_num_correct += torch.sum(preds == labels.data)
        print('test accuracy: {:4f}'.format(running_num_correct.double()/ dataset_sizes[phase]))

File: scripts/bird_species.R
library(torch)
library(torchvision)
library(torchdatasets)

library(dplyr)
library(pins)
library(ggplot2)

device <- if (cuda_is_available()) torch_device("cuda:0") else "cpu"

train_transforms <- function(img) {
  img %>%
    transform_to_tensor() %>%
    (function(x)
      x$to(device = device)) %>%
    transform_random_resized_crop(size = c(224, 224)) %>%
    transform_color_jitter() %>%
    transform_random_horizontal_flip() %>%
    transform_normalize(mean = c(0.485, 0.456, 0.406),
                        std = c(0.229, 0.224, 0.225))
}

valid_transforms <- function(img) {
  img %>%
    transform_to_tensor() %>%
    (function(x)
      x$to(device = device)) %>%
    transform_resize(256) %>%
    transform_center_crop(224) %>%
    transform_normalize(mean = c(0.485, 0.456, 0.406),
                        std = c(0.229, 0.224, 0.225))
}

test_transforms <- valid_transforms

train_ds <-
  bird_species_dataset("data", download = TRUE, transform = train_transforms)
valid_ds <-
  bird_species_dataset("data", split = "valid", transform = valid_transforms)
test_ds <-
  bird_species_dataset("data", split = "test", transform = test_transforms)

train_ds$.length()
valid_ds$.length()
test_ds$.length()

class_names <- train_ds$classes
length(class_names)

batch_size <- 16
train_dl <-
  dataloader(train_ds, batch_size = batch_size, shuffle = TRUE)
valid_dl <- dataloader(valid_ds, batch_size = batch_size)
test_dl <- dataloader(test_ds, batch_size = batch_size)

train_dl$.length()
valid_dl$.length()
test_dl$.length()

batch <- test_dl$.iter()$.next()
batch[[1]]$size()
batch[[2]]$size()

classes <- batch[[2]]
images <- as_array(batch[[1]]$cpu()) %>%
  aperm(perm = c(1, 3, 4, 2))
mean <- c(0.485, 0.456, 0.406)
std <- c(0.229, 0.224, 0.225)
images <- std * images + mean
images <- images * 255
images[images > 255] <- 255
images[images < 0] <- 0

par(mfcol = c(4, 4), mar = rep(1, 4))

images %>%
  purrr::array_tree(1) %>%
  purrr::set_names(class_names[as_array(classes)]) %>%
  purrr::map(as.raster, max = 255) %>%
  purrr::iwalk( ~ {
    plot(.x)
    title(.y)
  })

model <- model_resnet18(pretrained = TRUE)
model$parameters %>% purrr::walk(function(param) param$requires_grad_(FALSE))

num_features <- model$fc$in_features

model$fc <- nn_linear(in_features = num_features, out_features = length(class_names))


model <- model$to(device = device)

criterion <- nn_cross_entropy_loss()

optimizer <- optim_sgd(model$parameters, lr = 0.1, momentum = 0.9)

# find initial learning rate
# ported from: https://sgugger.github.io/how-do-you-find-a-good-learning-rate.html

losses <- c()
log_lrs <- c()

find_lr <-
  function(init_value = 1e-4,
           final_value = 10,
           beta = 0.98) {
    num <- train_dl$.length()
    mult = (final_value / init_value) ^ (1 / num)
    lr <- init_value
    optimizer$param_groups[[1]]$lr <- lr
    avg_loss <- 0
    best_loss <- 0
    batch_num <- 0

    for (b in enumerate(train_dl)) {
      batch_num <- batch_num + 1
      optimizer$zero_grad()
      output <- model(b[[1]]$to(device = device))
      loss <- criterion(output, b[[2]]$to(device = device))

      #Compute the smoothed loss
      avg_loss <- beta * avg_loss + (1 - beta) * loss$item()
      smoothed_loss <- avg_loss / (1 - beta ^ batch_num)
      #Stop if the loss is exploding
      if (batch_num > 1 && smoothed_loss > 4 * best_loss)
        break
      #Record the best loss
      if (smoothed_loss < best_loss || batch_num == 1)
        best_loss <- smoothed_loss

      #Store the values
      losses <<- c(losses, smoothed_loss)
      log_lrs <<- c(log_lrs, (log(lr, 10)))

      loss$backward()
      optimizer$step()

      #Update the lr for the next step
      lr <- lr * mult
      optimizer$param_groups[[1]]$lr <- lr
    }
  }

find_lr()

losses <- readRDS("scripts/losses.rds")
log_lrs <- readRDS("scripts/log_lrs.rds")

df <- data.frame(log_lrs = log_lrs, losses = losses)
ggplot(df, aes(log_lrs, losses)) +
  geom_point(size = 1) +
  theme_classic()

model <- model_resnet18(pretrained = TRUE)
model$parameters %>% purrr::walk(function(param)
  param$requires_grad_(FALSE))

num_features <- model$fc$in_features

model$fc <-
  nn_linear(in_features = num_features,
            out_features = length(class_names))


model <- model$to(device = device)

criterion <- nn_cross_entropy_loss()

optimizer <- optim_sgd(model$parameters, lr = 0.05, momentum = 0.9)

num_epochs <- 10

scheduler <- optimizer %>%
  lr_one_cycle(
    max_lr = 0.05,
    epochs = num_epochs,
    steps_per_epoch = train_dl$.length()
  )

batch_lrs <- c()

train_batch <- function(b) {

  optimizer$zero_grad()
  output <- model(b[[1]])
  loss <- criterion(output, b[[2]]$to(device = device))
  loss$backward()
  optimizer$step()
  scheduler$step()

  batch_lrs <<- c(batch_lrs, optimizer$param_groups[[1]]$lr)
  loss$item()

}

valid_batch <- function(b) {

  output <- model(b[[1]])
  loss <- criterion(output, b[[2]]$to(device = device))
  gc(full = TRUE)
  loss$item()
}

batch_lrs <- c()

for (epoch in 1:num_epochs) {

  model$train()
  train_losses <- c()
  it <- 0

  for (b in enumerate(train_dl)) {
    it <- it + 1
    loss <- train_batch(b)
    train_losses <- c(train_losses, loss)
  }

  model$eval()
  valid_losses <- c()

  for (b in enumerate(valid_dl)) {
    loss <- valid_batch(b)
    valid_losses <- c(valid_losses, loss)
  }

  cat(sprintf("\nLoss at epoch %d: training: %3f, validation: %3f\n", epoch, mean(train_losses), mean(valid_losses)))
}

df <- data.frame(x = 1:length(batch_lrs), y = batch_lrs)
df %>% ggplot(aes(x, y)) + geom_point(size = 1) +
  xlab("batch") +
  ylab("learning rate") +
  ggtitle("Learning rates (complete training)") +
  theme_classic()

model$eval()

model$eval()

test_batch <- function(b) {

  output <- model(b[[1]])
  labels <- b[[2]]$to(device = device)
  loss <- criterion(output, labels)
  test_losses <<- c(test_losses, loss$item())
  # torch_max returns a list, with position 1 containing the values
  # and position 2 containing the respective indices
  predicted <- torch_max(output$data(), dim = 2)[[2]]
  total <<- total + labels$size(1)
  # add number of correct classifications in this batch to the aggregate
  correct <<- correct + (predicted == labels)$sum()$item()
  gc(full = TRUE)

}

test_losses <- c()
total <- 0
correct <- 0

for (b in enumerate(test_dl)) {
  test_batch(b)
}


mean(test_losses)
test_accuracy <-  correct / total
test_accuracy
